### PR TITLE
Distributed example on C++ API (LibTorch) 

### PR DESCRIPTION
**Pull request by soumyadipghosh**
This PR adds an example involving distributed training using MPI on the C++ frontend similar to DistributedDataParallel in Python. This topic was raised in the forums [here](https://discuss.pytorch.org/t/libtorch-c-mpi-example/39671). Right now, this code is CPU only. 

Please let me know if this PR can be a worthwhile contribution.

cc @yf225 @pietern

### CODE REVIEW

**Comment by yf225** [comment]
@glaringlee @mrshenli Curious would this be a good addition? Thanks!

**Comment by glaringlee** [comment]
@yf225 This is good to me. I saw there is only 1 check (Run Examples) within this repo, is that as designed?
@mrshenli can you check the logic here, it seems fine to me though.

File: cpp/distributed/dist-mnist.cpp
Comment (mrshenli) [inline cpp/distributed/dist-mnist.cpp:116]: debug vestige?

File: cpp/distributed/dist-mnist.cpp
Comment (mrshenli) [inline cpp/distributed/dist-mnist.cpp:130]: could you please use clang-format to format code?

File: cpp/distributed/dist-mnist.cpp
Comment (mrshenli) [inline cpp/distributed/dist-mnist.cpp:63]: Should we make this an input arg instead of hard-coding a path?

File: cpp/distributed/dist-mnist.cpp
Comment (mrshenli) [inline cpp/distributed/dist-mnist.cpp:129]: cc @agolynski, have we exposed ProcessGroup C++ APIs? Can users call allreduce directly from process group instances?

File: cpp/distributed/dist-mnist.cpp
Comment (mrshenli) [inline cpp/distributed/dist-mnist.cpp:132]: Curious, is this MPI op blocking? Will it help speed up training if we launch all asynchronously and then wait for all?

Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp:132]: This MPI op is blocking. If we want to do it asynchronously, we can use [MPI_Iallreduce](https://www.mpich.org/static/docs/v3.1/www3/MPI_Iallreduce.html). However, that would just overlap the computation with communication. We still need to use [MPI_Wait](https://www.mpich.org/static/docs/v3.1/www3/MPI_Wait.html) to ensure that the non-blocking operation has completed before training moves to the next epoch.

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp:63]: Forgot to remove that! Should I put the path as a user input or search for it in a subfolder as shown in this [example](https://github.com/pytorch/examples/blob/master/cpp/mnist/mnist.cpp) ?

**Comment by glaringlee** [comment]
@yf225 @mrshenli 
I see there is a CI like test in this repo which run all the examples, any change that fail the execution will be detected in the test. For silence failure (logic change but won't stop the execution), can we add assert within the example so it breaks if any bc breaking change is made?

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp]: yeah, removed now.

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp]: done!

**Comment by mrshenli** [comment]
Do we need to add this new test to some `.sh` file, or will the CI automatically detect and include all new tests?

**Comment by glaringlee** [comment]
Should add dist-mnist after this line I think and do proper cleanup:
https://github.com/pytorch/examples/blob/master/run_python_examples.sh#L189
cc @yf225 @mrshenli @soumyadipghosh

**Comment by soumyadipghosh** [comment]
Also, for running tests, an MPI installation is necessary. How will that be added to the CI pipeline?

**Comment by glaringlee** [comment]
I think the CI is totally controlled by run_python_examples.sh. MPI can be installed when installing dependencies.
cc @seemethere

**Comment by soumyadipghosh** [comment]
cc @yf225 @glaringlee @mrshenli Can someone help me with the next steps ? I am not sure how to add the dependencies and the tests.

**Comment by glaringlee** [comment]
@malfet Are you familiar with this repo? Should we update run_python_examples.sh (add this test in) to make this test running on each code update? cc @mrshenli

**Comment by soumyadipghosh** [comment]
Can someone please help me with the next steps? I think the [run_python_examples.sh](https://github.com/pytorch/examples/blob/master/run_python_examples.sh) is only relevant for the python examples. It would be better to create something like run_cpp_examples.sh to test the cpp examples including this one.
cc @yf225 @glaringlee @mrshenli @malfet @seemethere @soumith

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp]: changed to "../data"

File: cpp/distributed/dist-mnist.cpp
Comment (osalpekar) [inline cpp/distributed/dist-mnist.cpp:119]: Can we create an MPI Process Group and call the process group `all_reduce` function instead?

File: cpp/distributed/dist-mnist.cpp
Comment (osalpekar) [inline cpp/distributed/dist-mnist.cpp:117]: I'm wondering whether we should add a comment here indicating that while this approach will allow synchronous data parallel training, it will lag behind DDP in performance since it synchronizes parameters after the backward pass whereas DDP overlaps synchronizing parameters and computing gradients in the backwards pass.

**Review by osalpekar — COMMENTED** [review]
Thanks for giving an update on your work in the PyTorch forums! Adding some thoughts that might improve this example

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp:117]: @osalpekar I can add that comment. Also I can call MPI_Iallreduce before the backward pass which is a non-blocking call, then call MPI_Wait for that call after the backward pass. This will ensure that backward pass and synchronizing parameters are overlapped. What do you say?

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp:119]: @osalpekar We can do that but will that help in some way?

File: cpp/distributed/dist-mnist.cpp
Comment (osalpekar) [reply cpp/distributed/dist-mnist.cpp:117]: I don't think that will have the same effect here. `loss.backward` is what actually computes the gradients so I don't think we can launch the synchronization collective before we've even started computing gradients.

DDP has a notion of buckets. We've instrumented the backward pass in DDP such that, once a bucket has been filled with gradients, we launch the allreduce. While the allreduce call is executing, we continue computing the next gradients to fill the next bucket, and so on. Trying to re-create this behavior efficiently is probably beyond the scope of this example so we can just stick to a blocking allreduce after the backward pass, but just in case somebody tries to measure the performance, we should include the comment so that they know this approach will be slower than DDP.

File: cpp/distributed/dist-mnist.cpp
Comment (osalpekar) [reply cpp/distributed/dist-mnist.cpp:119]: @soumyadipghosh good question - most of the reasons below exist for Python - let's check with @agolynski about ProcessGroup C++ API just to make sure this is allowed.

Using ProcessGroup is the preferred way of launching collectives and communication operations in PyTorch distributed. This way you can create a ProcessGroup using any underlying communication library you prefer (MPI, NCCL, Gloo, etc.), and simply call any of the collectives (allreduce, broadcast, etc.). Not only are you not tied to any comm library at the user program level, but you also gain some other benefits depending on the specific ProcessGroup type being used (error handling, timeout checking, perf optimizations, planned support for async Futures, etc.).

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp:117]: I see! Added the comment :)

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp:123]: @osalpekar Any update on how to proceed about the ProcessGroup C++ API?

**Comment by soumyadipghosh** [comment]
@osalpekar @mrshenli Since we haven't heard back about the ProcessGroup C++ APIs (see comments above), I created a separate file that involves NCCL for GPU communication that is somewhat based on the NCCL Example 2 [here](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/examples.html). I am not sure how to combine the CPU and GPU versions into one file because the NCCL routines will lead to errors when run on CPUs. Any ideas on the next steps?

File: cpp/distributed/dist-mnist.cpp
Comment (osalpekar) [reply cpp/distributed/dist-mnist.cpp:123]: @soumyadipghosh Apologies for the delay! [Here's](https://github.com/pytorch/pytorch/blob/c115957df086fd795abaae18c6b9bc1e80928a18/torch/lib/c10d/ProcessGroupMPI.hpp) a link to the ProcessGroupMPI header file. You can simply construct the ProcessGroup and then call any collectives on that process group (which are implemented as class methods - for example, allreduce). If the MPI you are using is CUDA-aware, you can use CUDA tensors.

We also have a [NCCL ProcessGroup](https://github.com/pytorch/pytorch/blob/c115957df086fd795abaae18c6b9bc1e80928a18/torch/lib/c10d/ProcessGroupNCCL.hpp) that you can as above for using the NCCL backend. This way you will not have to call into the NCCL functions for initialization, creating communicators, etc. - it is all done for you, so you can simply create the ProcessGroup and then use the provided collectives (for example, `ProcessGroupNCCL::allreduce`). 

If you want to use both CPUs and GPUs while training, the best option would be Gloo (we also have a ProcessGroupGloo in the same folder as the above 2) and another option is to use MPI if your version of MPI is CUDA-aware.

**Comment by osalpekar** [comment]
> @osalpekar @mrshenli Since we haven't heard back about the ProcessGroup C++ APIs (see comments above), I created a separate file that involves NCCL for GPU communication that is somewhat based on the NCCL Example 2 [here](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/examples.html). I am not sure how to combine the CPU and GPU versions into one file because the NCCL routines will lead to errors when run on CPUs. Any ideas on the next steps?

I added a comment above about the ProcessGroup C++ APIs - I think these should significantly simplify some of the initialization boilerplate that's here for NCCL. For the sake of this example, adding support for multiple ProcessGroups might be a bit too much since these examples should demonstrate to users how to efficiently use PyTorch Distributed training - we can stick to MPI for this example and include a comment that NCCL or Gloo can be used as alternatives. I think we should be in good shape to merge this in once we move to the MPI ProcessGroup.

We also have an ongoing discussion about native DDP C++ support (https://github.com/pytorch/pytorch/issues/48959), so please feel free to contribute ideas there as well!

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [inline cpp/distributed/dist-mnist.cpp:1]: this line leads to error!

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [inline cpp/distributed/dist-mnist.cpp:44]: Not sure if the above is suitable to initialize the MPI environment and create the ProcessGroup API

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [inline cpp/distributed/dist-mnist.cpp:88]: Since we are not using MPI directly, how to call an MPI primitive such as MPI_Wtime() separately in the code to measure the training time?

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [inline cpp/distributed/dist-mnist.cpp:118]: How to add the AllreduceOptions for the SUM operator?

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [inline cpp/distributed/dist-mnist.cpp:185]: Not sure if this is the right way to finalize the MPI environment?

**Comment by soumyadipghosh** [comment]
@osalpekar I agree that we can just stick to MPI for this example and so I removed the NCCL code for now. Regarding the move to ProcessGroupMPI, I made some initial changes to the code based on the c10d example [here](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/example/allreduce.cpp) but it doesn't compile! I have a lot of questions which I put as comments in the code, can you please clarify them?

File: cpp/distributed/dist-mnist.cpp
Comment (lasagnaphil) [reply cpp/distributed/dist-mnist.cpp:1]: Maybe is it related to the fact that the default PyTorch installation excludes support for MPI? It says that you need to recompile PyTorch from scratch if you want to use the MPI backend for torch.distributed. (https://pytorch.org/tutorials/intermediate/dist_tuto.html#communication-backends) As I'm also having this exact same issue, I'll try to do that tomorrow.

File: cpp/distributed/dist-mnist.cpp
Comment (osalpekar) [reply cpp/distributed/dist-mnist.cpp:44]: I think `ProcessGroupMPI::createProcessGroupMPI` would probably encompass all the initialization. It calls `initMPIOnce` for you and returns a correctly instantiated ProcessGroup

File: cpp/distributed/dist-mnist.cpp
Comment (osalpekar) [reply cpp/distributed/dist-mnist.cpp:88]: Since this is an illustrative example for future users, let's leave out timing code. For your personal use, you can still call `MPI_Wtime()` before and after whichever function you want to measure the time of right? For example, before and after `pg.allreduce`?

File: cpp/distributed/dist-mnist.cpp
Comment (osalpekar) [reply cpp/distributed/dist-mnist.cpp:118]: `AllreduceOptions` is a second optional argument for the `allreduce` function.
You can create the AllreduceOptions struct and set the `reduceOp = ReduceOp::SUM`, but this is already the default so you should be able to leave this as is.

File: cpp/distributed/dist-mnist.cpp
Comment (osalpekar) [reply cpp/distributed/dist-mnist.cpp:185]: `mpiExit` is called for you, you don't need to call it separately.

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp]: Ok done.

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp]: I was measuring the time for training. But for the sake of simplicity, I removed it.

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp]: Alright. Also adding the waiting mechanism as stated in [https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/test/ProcessGroupMPITest.cpp](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/test/ProcessGroupMPITest.cpp)

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp]: Removed it

File: cpp/distributed/dist-mnist.cpp
Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp:1]: @osalpekar How can this be fixed? I am not sure if @lasagnaphil ' s idea will work since Libtorch doesn't need to be compiled?

Comment (lasagnaphil) [reply cpp/distributed/dist-mnist.cpp:1]: The libtorch installation files that pytorch.org distributes are all precompiled libraries (with their accompanying headers). Chances are that even if you use the system's package manager (such as the Arch Linux AUR), they will use those official precompiled libraries. And from what I've seen, these do not include any support for MPI (for example, it's missing the header ProcessGroupMPI.hpp)

It seems that the build servers for pytorch.org do not include MPI support, which is understandable because there are multiple MPI implementations (OpenMPI, MPICH, MVAPICH, ...) and it's impractical to build and distribute the library separately for each library. So the only option would be to build and install libtorch from scratch. Also, the CMake script for c10d seems to include MPI support for the library only if a valid MPI installation is found (https://github.com/pytorch/pytorch/blob/c215ffb6a2de1c85a62a146a9e282908335ba3ed/torch/lib/c10d/CMakeLists.txt#L146), so you need to make sure an MPI installation exists on your system before calling ``python setup.py install``.

Comment (beniz) [reply cpp/distributed/dist-mnist.cpp:1]: As a side note, this might indicate that NCCL would make for a simpler example, though certainly less interesting in terms of showing how to deal with some more internals (e.g. intialization, ...).

Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp:1]: @lasagnaphil I see, thanks! Were you able to compile libtorch from scratch using any MPI?

Comment (lasagnaphil) [reply cpp/distributed/dist-mnist.cpp:1]: Compiling works now, as the headers (at least) are now properly installed on the system. However, I'm still having a linking error (``undefined reference to `c10d::ProcessGroupMPI::createProcessGroupMPI``) that I need to investigate. I tried to link the ``c10d`` CMake target to my executable but it says that it doesn't exist, which is a mystery.

Comment (lasagnaphil) [reply cpp/distributed/dist-mnist.cpp:1]: Ah, I figured it out. It seems like the file ``libc10d.a`` wasn't copied properly into the installation folder. Also, it seems that the particular static library needs to be linked directly with the exact path as the following

```
target_link_libraries(dist-mnist ${CMAKE_PREFIX_PATH}/lib/libc10d.a)
```
After this, and fixing a couple of more errors, I was able to successfully run the example (training seems to be doing all fine). Here's the pull request for the fix: https://github.com/soumyadipghosh/examples/pull/1

Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp:1]: @lasagnaphil Great! Can you share how you compiled libtorch from scratch using MPI?

Comment (lasagnaphil) [reply cpp/distributed/dist-mnist.cpp:1]: It took a while to figure out how to compile libtorch cleanly, I'll provide you a shell script: https://gist.github.com/lasagnaphil/3e0099816837318e8e8bcab7edcfd5d9

There are a lot of flags, and you might need to tweak some of them depending on your PC / usecase.

Take in mind that it takes quite some CPU time to compile! (For my Ryzen 7 2700X system, it took about 20~30 minutes)

Comment (osalpekar) [reply cpp/distributed/dist-mnist.cpp:1]: @soumyadipghosh The PR should be in good shape after @lasagnaphil's patch - we should be good to go ahead and approve. Are you still seeing build issues?

Comment (soumyadipghosh) [reply cpp/distributed/dist-mnist.cpp:1]: @osalpekar Sorry, I didn't have the chance to verify - will do it soon. For now, I have merged @lasagnaphil 's patch and added the link to the shell script for build in the Readme. We should be good to go ahead now. If I have build issues later, I will create an issue.

Comment (osalpekar) [reply cpp/distributed/dist-mnist.cpp:1]: @soumyadipghosh No problem, I'm accepting this for now. Would be great if we could verify the build before merging

**Review by osalpekar — APPROVED** [review]
Looks great - thanks for your work on this example @soumyadipghosh!

**Comment by soumyadipghosh** [comment]
@osalpekar Thanks for all your help! I have verified the build and everything works great! Ready for merging :)

**Comment by soumyadipghosh** [comment]
@osalpekar So how does this get merged?

**Pull request merged**
